subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/DefaultFileSystemLocationFingerprint.java

```diff
@@ -45,8 +50,73 @@ private static HashCode hashForType(FileType fileType, HashCode hash) {
         }
     }
 
+    @Override
+    public final void appendToHasher(Hasher hasher) {
+        hasher.putString(getNormalizedPath());
+        hasher.putHash(getNormalizedContentHash());
+    }
+
+    @Override
+    public FileType getType() {
+        return type;
+    }
+
     @Override
     public String getNormalizedPath() {
         return normalizedPath;
     }
+
+    @Override
+    public HashCode getNormalizedContentHash() {
+        return normalizedContentHash;
+    }
+
+    @Override
+    public final int compareTo(FileSystemLocationFingerprint o) {
+        int result = getNormalizedPath().compareTo(o.getNormalizedPath());
+        if (result == 0) {
+            result = getNormalizedContentHash().compareTo(o.getNormalizedContentHash());
+        }
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        DefaultFileSystemLocationFingerprint that = (DefaultFileSystemLocationFingerprint) o;
+
+        if (type != that.type) {
+            return false;
+        }
+        if (!normalizedContentHash.equals(that.normalizedContentHash)) {
+            return false;
+        }
+        return normalizedPath.equals(that.normalizedPath);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = type.hashCode();
+        result = 31 * result + normalizedContentHash.hashCode();
+        result = 31 * result + normalizedPath.hashCode();
+        return result;
+    }
+
+    @Override
+    public final String toString() {
```

Comment:
Please consider extracting the nested type calculation logic, like:
```java
@Override
public final String toString() {
    return String.format("'%s' / %s", getNormalizedPath(), getTypeString());
}

private String getTypeString() {
    switch (type) {
        case Directory:
            return "DIR";
        case Missing:
            return "MISSING";
        default:
            return normalizedContentHash.toString();
    }
}
```

Reply:
👍